File: packages/oxlint-plugin-react-doctor/src/plugin/rules/security-scan/tenant-static-proxy-risk.ts
import { defineRule } from "../../utils/define-rule.js";
import { isServerRouteSourcePath } from "./utils/is-server-route-source-path.js";
import { scanByPattern } from "./utils/scan-by-pattern.js";

export const tenantStaticProxyRisk = defineRule({
  id: "tenant-static-proxy-risk",
  title: "Tenant-controlled static asset proxy",
  severity: "warn",
  impact: "security",
  confidence: "heuristic",
  fix: "local",
  recommendation:
    "Bind tenant identity to the trusted host or authenticated org, canonicalize after decoding, reject traversal, and never let one tenant choose another tenant's asset prefix.",
  // `params` near `fetch` matched every dynamic route handler; the tenant
  // identifier must sit inside the asset-fetch call's own argument list.
  // `(?<!\.)org` keeps `.org` domain literals from counting.
  scan: scanByPattern({
    shouldScan: (file) => isServerRouteSourcePath(file.relativePath),
    pattern:
      /\b(?:fetch|path\.join|getObject\w*|GetObjectCommand|getSignedUrl|createReadStream)\s*\([^;]{0,200}(?:\$\{[^}]{0,100}\b(?:tenant|subdomain|workspace|hostPattern|(?<!\.)organization(?:Id|Slug)?)\b|\b(?:tenant|subdomain|workspace)(?:Id|Slug|Name)?\b\s*[,)+\].])/i,
    message:
      "Route code appears to compose tenant or subdomain input into a static/CDN/object-store fetch path.",
  }),
});

File: native/oxlint/scans/tenant-static-proxy-risk.rs
use lazy_regex::{Lazy, Regex, lazy_regex};

use super::{
    ScanFinding, get_location_at_index::get_location_at_index, scan_content::ScanContent,
};

const MESSAGE: &str = "Route code appears to compose tenant or subdomain input into a static/CDN/object-store fetch path.";

static TENANT_STATIC_PROXY_PATTERN: Lazy<Regex> = lazy_regex!(
    r"(?is)(?-u:\b)(?:fetch|path\.join|getObject[A-Za-z0-9_]*|GetObjectCommand|getSignedUrl|createReadStream)\s*\([^;]{0,200}(?:\$\{(?:[^}]{0,100}(?-u:\b)(?:tenant|subdomain|workspace|hostPattern)(?-u:\b)|organization(?:Id|Slug)?(?-u:\b)|[^}]{0,99}[^.A-Za-z0-9_]organization(?:Id|Slug)?(?-u:\b))|(?-u:\b)(?:tenant|subdomain|workspace)(?:Id|Slug|Name)?(?-u:\b)\s*[,)+\].])"
);

pub fn scan(relative_path: &str, source: &ScanContent<'_>) -> Vec<ScanFinding> {
    if !super::security_file_path::is_server_route_source_path(relative_path) {
        return Vec::new();
    }
    let normalized = source.normalized_scannable(false);
    let Some(found) = TENANT_STATIC_PROXY_PATTERN.find(normalized.as_ref()) else {
        return Vec::new();
    };
    let (line, column) = get_location_at_index(source, normalized.as_ref(), found.start());
    vec![ScanFinding::inherited(MESSAGE, line, column)]
}
